enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,307 @@ package hit_tracker_client
+			prometheus.Labels{
+				metrics.GroupID:              string(groupID),
+				metrics.EnqueueUpdateOutcome: "dropped_too_many_updates",
+			}).Add(float64(1))
+		return
+	}
+	metrics.RemoteHitTrackerUpdates.With(
+		prometheus.Labels{
+			metrics.GroupID:              string(groupID),
+			metrics.EnqueueUpdateOutcome: "enqueued",
+		}).Add(float64(1))
+
+	// Store the latest auth headers for this group for use in the async RPC.
+	authHeaders := authutil.GetAuthHeaders(ctx, h.authenticator)
+	groupHits.authHeaders = authHeaders
+
+	groupHits.hits = append(groupHits.hits, hit)
+}
+
 type TransferTimer struct {
+	ctx             context.Context
+	enqueueFn       func(context.Context, *repb.RequestMetadata, *hitpb.CacheHit)
+	invocationID    string
+	requestMetadata *repb.RequestMetadata
+	digest          *repb.Digest
+	start           time.Time
+	client          hitpb.HitTrackerServiceClient
 }
 
 func (t *TransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
+	hit := &hitpb.CacheHit{
+		RequestMetadata: t.requestMetadata,
+		Resource: &rspb.ResourceName{
+			Digest:    t.digest,
+			CacheType: rspb.CacheType_CAS,
+		},
+		SizeBytes: bytesTransferredClient,
+		Duration:  durationpb.New(time.Since(t.start)),
+	}
+	t.enqueueFn(t.ctx, t.requestMetadata, hit)
 	return nil
 }
 
 func (t *TransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
-	return nil
+	return status.InternalError("Unxpected call to hit_tracker_client.Record()")
+}
+
+func (h *HitTrackerClient) TrackDownload(digest *repb.Digest) interfaces.TransferTimer {
+	return &TransferTimer{
+		ctx:             h.ctx,
+		enqueueFn:       h.enqueueFn,
+		requestMetadata: h.requestMetadata,
+		digest:          digest,
+		start:           time.Now(),
+		client:          h.client,
+	}
 }
 
-func (h *HitTracker) TrackDownload(d *repb.Digest) interfaces.TransferTimer {
-	return &TransferTimer{}
+// Writes hit the backing cache, so no need to report on hit-tracking.
+func (h *HitTrackerClient) TrackUpload(digest *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h *HitTracker) TrackUpload(d *repb.Digest) interfaces.TransferTimer {
-	return &TransferTimer{}
+func (h *HitTrackerFactory) runWorker() {
+	for {
+		select {
+		case <-h.quit:
+			h.stopped.Store(false)
+			return
+		case <-h.ticker:
+			h.sendTrackRequests(context.Background())
+		}
+	}
+}
+
+// Sends one round of HitTrackerService.Track requests from the queue. This
+// funciton will send one RPC per group that has any pending hit-updates, and
+// return the number of hit-updates sent.
+func (h *HitTrackerFactory) sendTrackRequests(ctx context.Context) int {
+	i := 0
+	groups := make([]groupID, len(h.hitsByGroup))
+	h.mu.Lock()
+	for group := range h.hitsByGroup {
+		groups[i] = group
+		i++
+	}
+	h.mu.Unlock()
+
+	for _, groupID := range groups {
```

Comment:
sorry, I should have pointed this out before - with the current implementation, I think you'll have a bunch of workers that are prematurely flushing batches, since each worker is independently waiting on the ticker. We'll probably frequently run into cases where one worker has just finished flushing the batch for a particular group ID, then a single request gets added to the batch, then a second worker looks at that group ID again (as part of its flush loop), then flushes just that single request.

One way to work around this would be to record the timestamp at which the first request was added to the batch, and don't flush if the time elapsed since then is less than 1 ticker period.

Reply:
Changed to use a queue as we discussed offline.